ui_update: Show the number of messages to be moved.

whilstsomebody opened the pull request:
This PR adds a message count display in the **Move topic** and **Move messages** modals, showing the number of messages that will be affected by the move action. The information is displayed below the stream/topic or selection specification, depending on the context of the modal.

**Tasks completed**

- [x] Added message count display below the stream/topic in the "Move topic" UI.
- [x] Added message count display below the selection area in the "Move messages" UI.
- [x] The count will display in the format: "N messages will be moved."
- [x] Special handling for N = 1.
- [x] Handling special case when the message view is not topic narrowed. 

## Screenshots

### Move messages modal


<table>
<thead>
<tr>
<th>Before</th>
<th> After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<img src="https://github.com/user-attachments/assets/52a9205e-1dde-413a-ab12-20fd3af12b4a" alt="Before Image 1">
</td>
<td>
<img src="https://github.com/user-attachments/assets/5b442d2d-d543-48fa-bdda-2b9b6d7b8c5b" alt="After Image 1">
</td>
</tr>
<tr>
<td>
<img src="https://github.com/user-attachments/assets/177decf5-c9df-4b23-8eac-9280f0e413fc" alt="Before Image 2">
</td>
<td>
<img src="https://github.com/user-attachments/assets/6c800d47-357d-4702-8129-03238359687e" alt="After Image 2">
</td>
</tr>
</tbody>
</table>

### Move topic modal


<table>
<thead>
<tr>
<th>Before</th>
<th> After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<img src="https://github.com/user-attachments/assets/33242a5d-989e-4550-a5bd-4dfefa8ca9bb" alt="Before Image 1">
</td>
<td>
<img src="https://github.com/user-attachments/assets/baf78b21-a56c-4f24-b5fd-fedd535d4679" alt="After Image 1">
</td>
</tr>
<tr>
<td>
<img src="https://github.com/user-attachments/assets/8838b184-c2a1-43fa-9a5b-e1225c172d7e" alt="Before Image 2">
</td>
<td>
<img src="https://github.com/user-attachments/assets/8a56a72f-2e63-4674-bc8d-ef905c7fe3c8" alt="After Image 2">
</td>
</tr>
</tbody>
</table>

###  Corner Cases

<details>
<summary> Check for N = 1 and N > 1 </summary>
<table>
<thead>
<tr>
<th>For N = 1</th>
<th>For N > 1</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<img src="https://github.com/user-attachments/assets/1d7a2740-ba9c-464f-9265-52b1c73f0501" alt="Before Image 1">
</td>
<td>
<img src="https://github.com/user-attachments/assets/442936f7-b0d3-483c-a3ac-ceeab28d2e53" alt="After Image 1">
</td>
</tr>
</tbody>
</table>
</details>

<details>
<summary> Is view topic narrowed? </summary>
<table>
<thead>
<tr>
<th>True</th>
<th>False</th>
</tr>
</thead>
<tbody>
<tr>
<td>

https://github.com/user-attachments/assets/182404e7-9a2f-4cac-8ca8-fdf6c25162ce


</td>
<td>

https://github.com/user-attachments/assets/c1609203-e299-4afc-81e3-4622f7a7f47a


</td>
</tr>
</tbody>
</table>
</details>



Fixes: #23115 .

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

whilstsomebody commented:
Hey @alya , the PR is ready for review. Can you please have a look?
Thanks!

whilstsomebody commented:
Hey @alya , the PR is ready for a review. Can you take a look?
Thanks!

whilstsomebody commented:
I have updated the design of the confirmation modal in agreement with the [CZO](https://chat.zulip.org/#narrow/channel/101-design/topic/Design.20of.20Message.20counter.20in.20confirmation.20modals).

whilstsomebody commented:
Hey @alya, the PR is ready for review. Can you take a look?
Thanks.

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -619,6 +620,56 @@ export async function build_move_topic_to_stream_popover(
+                    defaultMessage:
+                        "{count, plural, one {# message} other {# messages}} will be moved.",
+                },
+                {count: message_move_count},
```
I guess this code should be in the `post_render` function.

Also, would be good to not re-use `default_message_move_count` as more messages can be received while user changes the option. So, instead would be good to just create a function which retruns count based on the selected option which can be used for both the cases - while setting the text initially and changing the text with options.

sahil839 reviewed (commented):
Haven't tested yet but did some initial code review.

whilstsomebody commented on web/src/message_util.ts:
```diff
@@ -101,6 +101,27 @@ export function get_messages_in_dm_conversations(user_ids_strings: Set<string>):
+
+        if (has_found_the_current_message) {
+            messages.push(msg);
+        }
```
What about this name - `get_current_and_following_messages_in_topic`?

whilstsomebody replied:
Is it okay if I leave the function `get_messages_in_topic_sent_after_current_message`  to return an array of messages instead of the length? I feel it would be consistent with the surrounding code.

whilstsomebody commented:
Hey @sahil839 , I have made the changes as per your suggestions. The PR is ready for another review.

whilstsomebody commented:
Hey @sahil839 , I have made the changes as per your suggestions. The PR is ready for another review.
Thanks !

timabbott commented on web/src/stream_popover.js:
```diff
@@ -597,6 +612,48 @@ export async function build_move_topic_to_stream_popover(
+                current_stream_id,
+                topic_name,
+            ).length;
+        }
```
It seems fragile that this conditional needs to exactly match some other conditional for what the initial value is; can you align this to the other code?

timabbott replied:
Ideally you'd just call `get_count_of_messages_to_be_moved` here as well...

whilstsomebody commented:
Hey @timabbott , on the basis of my understanding of your review I have made some changes. Please take a look.

alya commented:
@sahil839 tagging as being reviewed by you.

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -552,6 +553,20 @@ export async function build_move_topic_to_stream_popover(
+    function get_count_of_messages_to_be_moved(selected_option) {
+        if (selected_option === "change_one") {
+            return 1;
+        } else if (selected_option === "change_later") {
```
No need of `else if` here, just `if` should work since the previous block returns.

sahil839 commented on web/src/message_util.ts:
```diff
@@ -84,6 +84,23 @@ export function add_new_messages_data(
+    topic: string,
+    messaged_id: number,
+): number {
+    const all_messages = get_messages_in_topic(stream_id, topic);
```
Would be good to directly use `.filter().length` I guess.

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -597,6 +612,35 @@ export async function build_move_topic_to_stream_popover(
         });
+
+        const default_message_move_count = get_count_of_messages_to_be_moved(
+            args.from_message_actions_popover ? "change_later" : "change_all",
```
This can directly be computed from the selected value in the `get_count_of_messages_to_be_moved` function instead of passing this here.

whilstsomebody replied:
Hey @sahil839,

The main purpose behind this approach is that `args.from_message_actions_popover == true` indicates that I clicked on `move messages`, while the `else` case means I clicked on `move topic`. This boolean, `args.from_message_actions_popover`, differentiates between the two options. For `move messages`, the default is to "`move this and the following messages`," and for `move topic`, it should show the total count of messages in the topic.

I’m not sure if it's possible to handle both cases without using this boolean.

An alternative approach would be to introduce a boolean parameter in the `get_count_of_messages_to_be_moved` function.

Could you please advise which approach is better, or if I should consider another option?

sahil839 replied:
Ok, yes. I missed that we cannot change propagate mode while opening the modal from topic popover. So, the previous approach was fine I guess.

whilstsomebody replied:
Sure, I reversed the changes.
Thanks!

sahil839 commented on web/templates/move_topic_to_stream.hbs:
```diff
@@ -16,12 +16,13 @@
             <option value="change_all" {{#if (eq message_placement "first")}}selected{{/if}}> {{t "Move all messages in this topic" }}</option>
         </select>
         {{/if}}
+        <p id="move_messages_count" class="message_count" ></p>
```
Extra space after "message_count" can be removed to keep formatting consistent.

sahil839 commented on web/styles/popovers.css:
```diff
@@ -938,6 +938,10 @@ ul.popover-group-menu-member-list {
         }
+
+        .message_count {
+            margin: -10px 0 10px 2px;
```
I guess the bottom margin of 10px for the dropdown can be removed and then the top margin of -10px will not be needed here.

I am not sure about the 2px margin here. This does not look bad or something but the "Expires on" text in invite modal doesn't have any left margin.

whilstsomebody replied:
Thanks for the correction. Left margin is not required. But we need both top and bottom margin. 

If I remove both top and botton margin:

![image](https://github.com/user-attachments/assets/9f48f8b5-0d05-4f51-9b8a-449cbb8aafc1)

If I keep the top margin and remove the botton margin:

![image](https://github.com/user-attachments/assets/f4ac6561-ad14-4d8c-a5ee-5d7c96a64262)

If I keep both the margin:

![image](https://github.com/user-attachments/assets/d7a73edf-8da0-4826-a060-b9e7cb7ea8a3)


The 2nd and 3rd one look almost same but on closely observing we can see that the third one is better.

sahil839 replied:
I mentioned removing the bottom margin from the select dropdown which will then make removing the top margin for the count text possible.

whilstsomebody replied:
Hey @sahil839,
I looked everywhere and confirmed that the select dropdown is using default styling with no local styles applied. To achieve the required design, we’ll need to add some styling. I believe a top and bottom margin is necessary.
Please tell if I am mistaken.
Thanks!

sahil839 replied:
I can see [this CSS](https://github.com/zulip/zulip/blob/d448b751763e01ea6b52d943c61115a2507d17fa/web/styles/zulip.css#L938) being applied using the browser inspector.

whilstsomebody replied:
Okay, I see. 
Thank you for the help! 
I’m now pushing the remaining changes. Thanks again!

sahil839 reviewed (commented):
Posted a few comments.

whilstsomebody commented:
Hey @sahil839, 
I’ve addressed all your comments and made the changes as per your suggestions. The PR is ready for review again. Could you please take a look?
Thanks!

whilstsomebody commented:
Hey @sahil839 , I have addressed the remaining comments. Could you please take a look?
Thanks!

whilstsomebody commented:
Hey @sahil839,
I have addressed the remaining comments. If there’s anything else I can improve, please let me know.
Thanks!

whilstsomebody commented:
Hey @laurynmm , I have made some changes in the code. Could you take a look? 
Thanks!

whilstsomebody commented:
Hey @sahil839,
I have addressed the remaining comments.Please take a look.
If there’s anything else I can improve, please let me know.
Thanks!

timabbott commented on web/src/stream_popover.js:
```diff
@@ -552,6 +553,20 @@ export async function build_move_topic_to_stream_popover(
+            );
+        }
+        return message_util.get_messages_in_topic(current_stream_id, topic_name).length;
+    }
```
This logic is only correct if `messages_lists.current.data.fetch_status.has_found_newest/oldest` are both true, depending on the mode; i.e. we have all the messages in the the current topic locally cached. This will often be the case, but is not guaranteed to be.

So I think we need to either (1) ask the server if we don't know, which would likely be annoying/expensive with current APIs or (2) have a variant of the wording to say "At least {n} messages will be moved." if the client cannot show it has the complete set.

whilstsomebody replied:
Hey @timabbott, based on my understanding, I think the following approach should be okay:

If `messages_lists.current.data.fetch_status.has_found_newest/oldest` are both true, we can display "N messages will be moved." If not, we can show "At least N messages will be moved."

However, I'm concerned this might be ambiguous for users, as sometimes they will see "N messages" and other times "At least N messages."

sahil839 replied:
I guess we would also need to handle the case when the message view is not narrowed to a topic while moving message because in that case `has_found_newest/oldest` fields will not represent the data fetched for that particular topic. Not sure how to do it, maybe we can just show "Atleast N messages ..." for that cases as well.

whilstsomebody replied:
Hey @sahil839, could you suggest if we could limit ourselves to using only "At least N messages..."? I think this would be less ambiguous for users, as they would consistently see "At least N messages..." rather than sometimes seeing "N messages..." and other times "At least N messages...".

sahil839 replied:
I am not sure but I feel would be better to just show "N messages ... " when we are sure. @timabbott might provide more feedback.

whilstsomebody commented:
Hey @timabbott  @sahil839 , I have made some changes. Please take a look.

For the case when `messages_lists.current.data.fetch_status.has_found_newest/oldest` both are true,
we shall show "N messages will be moved" for both 2nd and 3rd select option.

![image](https://github.com/user-attachments/assets/e9d2977e-5e54-4854-a257-594f87666bfb)

![image](https://github.com/user-attachments/assets/d42c6dd0-b928-4d1c-b52f-57e09a9680d2)


For the case when atleast one of `messages_lists.current.data.fetch_status.has_found_newest/oldest` is false,
we shall show "Atleast N messages will be moved" for both 2nd and 3rd select option.

![image](https://github.com/user-attachments/assets/6b603fb3-4a49-47d4-b000-051e6946ab0d)

![image](https://github.com/user-attachments/assets/fd9b8049-a283-4547-b528-bf90949ac62e)

whilstsomebody commented:
Hey @sahil839, could you take a look at the latest changes I've made?

timabbott commented on web/src/stream_popover.js:
```diff
@@ -552,6 +553,60 @@ export async function build_move_topic_to_stream_popover(
+            );
+        } else if (
+            message_lists.current.data.fetch_status.has_found_newest &&
+            message_lists.current.data.fetch_status.has_found_oldest
```
These are functions! You need to actually call them. Please be more careful in reading the code you're using -- this can't possibly be correct.

whilstsomebody replied:
Done! @timabbott . Please take a look.

timabbott commented on web/src/message_util.ts:
```diff
@@ -84,6 +84,15 @@ export function add_new_messages_data(
+): number {
+    const all_messages = get_messages_in_topic(stream_id, topic);
+    return all_messages.filter((msg) => msg.id >= message_id).length;
+}
```
As noted in the chat.zulip.org discussion, this function (and its parallel code path) needs to be tweaked to also check if the message is already marked unread.

whilstsomebody replied:
@Tim Abbott, I’m a bit confused. Did you mean "already marked unread" or "already marked read"?

whilstsomebody replied:
Hey @timabbott , I am a bit confused in this portion. Could you explain me why we need to check if the message is already marked unread,

timabbott replied:
Oh, never mind, this is not about "mark as unread". Yes this is fine.

timabbott commented:
I think you need to do some debugging about why the test is failing. (Was it just a nondeterministic flake?). Reverting to obviously wrong code (checking the boolean value of whether the function exists, not what it returns is not a useful thing to try when you get an error -- you need to debug it. The docs have some techniques.

whilstsomebody commented:
> I think you need to do some debugging about why the test is failing. (Was it just a nondeterministic flake?). Reverting to obviously wrong code (checking the boolean value of whether the function exists, not what it returns is not a useful thing to try when you get an error -- you need to debug it. The docs have some techniques.

Earlier, I directly called the two functions in the conditional, but that caused a Puppeteer error. Interestingly, when I defined two variables for the same purpose and passed those variables to the conditional, the tests passed. I wonder why that happened. :)

That said, the PR is ready for another review. @sahil839, could you please take a look?

whilstsomebody commented:
> I think you need to do some debugging about why the test is failing. (Was it just a nondeterministic flake?). Reverting to obviously wrong code (checking the boolean value of whether the function exists, not what it returns is not a useful thing to try when you get an error -- you need to debug it. The docs have some techniques.

Earlier, I directly called the two functions in the conditional, but that caused a Puppeteer error. Interestingly, when I defined two variables for the same purpose and passed those variables to the conditional, the tests passed. I wonder why that happened. :)

That said, the PR is ready for another review. @sahil839, could you please take a look?

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -562,6 +563,51 @@ export async function build_move_topic_to_stream_popover(
+    function update_move_messages_count_text(selected_option) {
+        const message_move_count = get_count_of_messages_to_be_moved(selected_option);
+        const has_found_newest = message_lists.current.data.fetch_status.has_found_newest();
+        const has_found_oldest = message_lists.current.data.fetch_status.has_found_oldest();
```
This still does not handle the case when moving messages from a view which is not a topic narrow correctly as mentioned [here](https://github.com/zulip/zulip/pull/32155/files#r1837869498).

Also, I guess just checking `has_found_newest` would be enough when selected option is "change_later".

whilstsomebody commented:
Thanks for the review @sahil839 . I have updated the PR to align the code according to your instruction. Could you please take a look?

whilstsomebody commented on web/src/stream_popover.js:
```diff
@@ -562,6 +564,57 @@ export async function build_move_topic_to_stream_popover(
+            current_filter &&
+            current_filter._terms.length === 2 &&
+            current_filter._terms.some((term) => term.operator === "channel") &&
+            current_filter._terms.some((term) => term.operator === "topic");
```
Does it mean that the count should show the number of messages in the stream?

whilstsomebody commented:
Hey @sahil839 , I have updated the PR. Could you please take a look?

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -562,6 +564,57 @@ export async function build_move_topic_to_stream_popover(
+        const has_found_oldest = message_lists.current.data.fetch_status.has_found_oldest();
+
+        let message_text;
+        if (narrow_state.narrowed_by_topic_reply || narrow_state.narrowed_by_stream_reply) {
```
The functions here are not being called.

whilstsomebody replied:
Ohh sorry, I missed that.

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -562,6 +564,57 @@ export async function build_move_topic_to_stream_popover(
+            if (
+                selected_option === "change_one" ||
+                (selected_option === "change_later" && has_found_newest) ||
+                (selected_option === "change_all" && has_found_newest && has_found_oldest)
```
This is not correct. We want to show the "At least N messages will be moved text" when one of these is not true, but currently we would not show any text if the message view is narrowed to topic or stream but one of these is not true.

whilstsomebody commented:
Hey @sahil839 , I have made the required updates. Could you please take a look?

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -562,6 +564,59 @@ export async function build_move_topic_to_stream_popover(
+
+        if (
+            (is_topic_narrowed || is_stream_narrowed) &&
+            (selected_option === "change_one" ||
```
For `change_one` option, there will always be one message irrespective of the narrow, so that should be fixed.

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -562,6 +564,59 @@ export async function build_move_topic_to_stream_popover(
+        const has_found_newest = message_lists.current.data.fetch_status.has_found_newest();
+        const has_found_oldest = message_lists.current.data.fetch_status.has_found_oldest();
+        const is_topic_narrowed = narrow_state.narrowed_by_topic_reply();
+        const is_stream_narrowed = narrow_state.narrowed_by_stream_reply();
```
I think we would also need to check whether the view is narrowed to the same stream or topic for which message is being moved as we can also move topic from the left sidebar popover.

whilstsomebody replied:
Hey @sahil839, I want to confirm one thing. I need to add a check to verify if the messages are being moved to the same stream. If yes, then I should check whether the messages are being moved to the same topic. If both conditions are true, the count should display 0. In all other cases, it should display the count. Am I right?

whilstsomebody commented:
Could you guide me how to target the stream selection dropdown correclty?

whilstsomebody commented:
Hey @sahil839 , I have made the changes. Could you please take a look?

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -562,6 +564,64 @@ export async function build_move_topic_to_stream_popover(
+
+        let message_text;
+        if (
+            (is_same_stream || is_same_topic) &&
```
Don't think this logic is correct. Like for a case where stream is same but it is narrowed to a different topic, then the `message_lists.current` will not contain data of the topic from whihch message is being moved.

Please just go through the cases once and make sure that the logic is correct.

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -562,6 +564,64 @@ export async function build_move_topic_to_stream_popover(
+        const has_found_newest = message_lists.current.data.fetch_status.has_found_newest();
+        const has_found_oldest = message_lists.current.data.fetch_status.has_found_oldest();
+        const narrowed_stream_id = terms[0].operand;
+        const narrowed_topic_name = terms[1].operand;
```
There are already helper functions in `narrow_state` to get the current stream ID and topic name.

whilstsomebody commented:
Hey @sahil839 , I have made the changes mentioned by you. Could you please take a look?

sahil839 commented on web/src/stream_popover.js:
```diff
@@ -562,6 +564,64 @@ export async function build_move_topic_to_stream_popover(
+        let message_text;
+        if (
+            is_same_stream &&
+            is_same_topic &&
```
This is also not correct. `is_same_topic` would never be true for a stream narrow, but we can be sure of the messages if the newest and/or oldest message is fetched.

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -585,6 +586,70 @@ export async function build_move_topic_to_stream_popover(
+            (((is_same_stream && is_same_topic && (is_topic_narrowed || is_stream_narrowed)) ||
+                (has_found_newest && has_found_oldest)) &&
+                ((selected_option === "change_later" && has_found_newest) ||
+                    (selected_option === "change_all" && has_found_newest && has_found_oldest)))
```
This is difficult to read. I guess this can be written as -
```js
const can_have_exact_count_in_narrow = (is_stream_narrowed && is_same_stream) || (is_topic_narrowed && is_same_stream && is_same_topic);
let exact_message_count = false;
if (selected_option === "change_one") {
    exact_message_count = true;
} else if (selected_option === "change_all" && can_have_exact_count_in_narrow && has_found_newest && has_found_oldest) {
    exact_message_count = true;
} else if (selected_option === "change_later" && can_have_exact_count_in_narrow && has_found_newest) {
    exact_message_count = true;
}

// Then the code to set the message_text as per exact_message_count
```

sahil839 reviewed (commented):
Posted some comments.

whilstsomebody commented on web/src/.cph/.stream_popover.js_cc3e054829ad627743075c0cf2915159.prob:
```diff
@@ -0,0 +1 @@
+{"name":"Local: stream_popover","url":"/home/aman/zulip-2/zulip/web/src/stream_popover.js","tests":[{"id":1732537410794,"input":"","output":""}],"interactive":false,"memoryLimit":1024,"timeLimit":3000,"srcPath":"/home/aman/zulip-2/zulip/web/src/stream_popover.js","group":"local","local":true}
```
Oh, sorry! I had accidentally opened the CPH extension, which added this file, and I missed it.

whilstsomebody commented on web/src/stream_popover.ts:
```diff
@@ -585,6 +586,70 @@ export async function build_move_topic_to_stream_popover(
+            .all_rendered_message_lists()[0]
+            ?.data.fetch_status.has_found_newest();
+        const has_found_oldest = message_lists
+            .all_rendered_message_lists()[0]
```
Initially, I removed that due to the undefined behavior of `message_lists.current`, but since it's just a TypeScript error, I think it's fine now.

whilstsomebody commented:
Hey @sahil839 , I have addressed all the comments and have made the changes. Could you please take a look?

whilstsomebody commented:
Hey @sahil839, could you please review the changes I’ve made and let me know if there’s anything else I need to do for this PR?

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -585,6 +586,87 @@ export async function build_move_topic_to_stream_popover(
+            has_found_newest
+        ) {
+            exact_message_count = true;
+        }
```
This is still somewhat hard to read. Maybe writing it like this would help -
```js
if (selected_option === "change_one") {
    exact_message_count = true;
} else if (can_have_exact_count_in_narrow) {
    const has_found_newest = message_lists.current?.data.fetch_status.has_found_newest();
    const has_found_oldest = message_lists.current?.data.fetch_status.has_found_oldest();

    if (selected_option === "change_later" && has_found_newest) {
        exact_message_count = true;
    }

    if (selected_option === "change_all" && has_found_newest && has_found_oldest) {
        exact_message_count = true;
    }
}
```

whilstsomebody commented:
Hey @sahil839 , I have made the suggested changes. Could you place take a look?

whilstsomebody commented:
Hey @sahil839 , I have made the suggested changes. Could you place take a look?

sahil839 commented:
Looks good to me. @timabbott ready for your review.

timabbott commented on web/styles/zulip.css:
```diff
@@ -938,7 +938,6 @@ div.focused-message-list {
     width: auto;
-    margin-bottom: 10px;
     max-width: 100%;
 }
```
This should probably move to `web/styles/popovers.css` and also be on `.modal_select`, not `select`.

alya commented:
[Additional discussion on CZO](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.9B.A0.EF.B8.8F.20wrong.20message.20count.20in.20move-messages.20dialog)
